online-store

MOONcitizenX opened the pull request:
### 1. Task: https://github.com/rolling-scopes-school/tasks/tree/master/tasks/online-store-team  
  
###  2.  
![chrome_KdigiGYu5v](https://user-images.githubusercontent.com/104726176/211166254-23d96f10-7dd6-4c96-999b-4e0ad18f6bb8.png)



### 3. Deploy: https://mkoroleva5-online-store.netlify.app/  
  
### 4. 07.01.2023 / 10.01.2023

###  5.  My score 300 / 300:   

#### Страница товаров с фильтрами  
  
  1. Реализована фильтрация продуктов +40
   - [x] присутствуют два блока фильтров, по категории и брэнду/названию, где реализован список с возможностью выбрать конкретную категорию/брэнд. +5 за каждый блок `+10`
   - [x] присутствуют два блока фильтров с ползунками (dual-slider), по диапазонам цены и наличию на складе. Вы можете заменить наличие на складе на другое свойство товара выбранной Вами тематики. +5 за каждый блок `+10`
   - [x] изменение любых фильтров должно добавлять в query-параметры соответствующие данные, для того, чтобы при перезагрузке страницы восстанавливалось текущее состояние страницы с примененными фильтрами `+10`
  - [x] при применении любого фильтра, должны динамически пересчитываться все фильтры и меняться состояние элементов в них. Например, при выборе определенной категории товара, в других фильтрах должно пересчитываться количество найденных товаров с указанием сколько найдено при такой конфигурации фильтров, а также сколько доступно без учета текущего поиска. В случае с dual-slider должен меняться их состояние-диапазон. (см. демо) `+10`
2. Реализована сортировка продуктов +20
   - [x] присутствует возможность сортировки продуктов минимум по 2 критериям, например по цене и рейтингу (от наименьшего к большему и наоборот) `+10`
   - [x] данные сортировки (если применялась) должны быть добавлены в query-параметр адресной строки. При копировании ссылки и открытии в новом окне/перезагрузке страницы, сортировка должна применяться к найденным продуктам `+5`
   - [x] при открытии в новом окне ссылки, которая содержит query-параметр сортировки, блок сортировки должен менять состояние, которое указывает, какая сортировка выбрана `+5`
3. Реализован текстовый поиск по всем данным продуктов +15
   - [x] присутствует input для поиска-фильтрации продуктов по любым данным. Например, поиск должен работать по всем полям продукта (название, описание, цена, кол-во на складе и т.д.) за исключением не информативных полей (id, url картинки и тд.) `+5`
   - [x] работает одновременно с другими фильтрами `+5`
   - [x] добавляется в query-параметры и при перезагрузке страницы восстанавливается состояние текущего поиска. Input также восстанавливает состояние и содержит поисковый текст `+5`
4. Реализовано переключение вида найденных продуктов +10
   - [x] реализована возможность переключения вида найденных товаров. Например списком / блоком карточек и т.д. Не менее двух вариантов. `+5`
   - [x] добавляется в query-параметры. При перезагрузке страницы восстанавливается состояние текущего вида. Выбранный режим вида выделяется `+5`
5. Реализован роутинг с query-параметрами +10
   - [x] все примененные фильтра, сортировка, вид отображения продуктов, текстовый поиск содержаться в query-параметрах и при перезагрузке страницы восстанавливают её состояние `+10`
6. Реализованы кнопки сброса и копирования поиска +10
   - [x] реализована кнопка сброса всех фильтров `+3`
   - [x] реализована кнопка копирования в буфер памяти текущего поиска, то есть адресную строку `+4`
   - [x] кнопка копирования текущего поиска меняет состояние при копировании (как Вы это реализуете - решать Вам, можете сделать как в демо) `+3`
7. Реализован блок кол-ва найденных товаров +5
   - [x] после применение фильтров отображается кол-во найденных товаров `+3`
   - [x] в случае, если товаров не нашлось, информация об этом дополнительно выводится в блоке товаров `+2`.
8. Поведение карточек найденных товаров +10
   - [x] карточка товара содержит кнопку добавления в корзину. Состояние кнопки меняется при добавлении/удалении, а также восстанавливается если товар был добавлен на других страницах `+5`
   - [x] при клике на товар должен происходить переход на страницу с описанием товара `+5`  
   
#### Страница корзины товаров  +60
  
1. Реализован блок отображения добавленных продуктов +5
   - [x] продукты выводятся списком, у каждого продукта присутствуют полные данные (имя, категория, цена и т.д.) и блок для управления кол-вом данного товара `+5`
2. Реализовано увеличение кол-ва конкретного товара и его удаление +10
   - [x] у каждого товара есть кнопка увеличения его кол-ва в корзине. При увеличении кол-ва, также изменяется состояние кол-ва и общей цены в Header-e и на странице корзины товаров. А также увеличивается общая сумма этого товара у него в списке. При этом, нельзя добавить товара больше, чем есть на складе. Информация о наличии на складе также должна выводится `+5`
   - [x] у каждого товара есть кнопка уменьшения его кол-ва в корзине. Если при уменьшении кол-ва товара его значения становится менее 1, то товар автоматический удаляется из корзины и текущей страницы. Также изменяется состояние кол-ва и общей цены в Header-e и на странице корзины товаров `+5`
3. Реализована пагинация +15
   - [x] добавлена пагинация с выбором кол-ва товара на одной странице. При изменении кол-ва товаров на странице, автоматически пересчитывается кол-во страниц для переключения. Нельзя переключиться на несуществующую страницу. Также, при удалении товара, пагинация также должна пересчитываться в случае, если после удаления страниц стало меньше, чем та, на которой находился пользователь. Например, выбран лимит 3 товара на странице, всего в корзине 4 товара, соответственно имеем 2 страницы и если пользователь, находясь на второй странице(где есть 1 товар) удаляет его, то автоматический должна переключаться страница на 1 назад. Данные также добавляются в query-параметры, а состояние текущей выбранной страницы и лимита товаров на ней восстанавливается при перезагрузке страницы `+8`
   - [x] если товаров в корзине нет, то все блоки страницы скрываются и вместо них выводится соответствующее сообщение `+5`
   - [x] товар на каждой странице имеет порядковый номер с учетом того, сколько товаров есть до него на предыдущих страницах `+2`
4. Хранение данных в localStorage +10
   - [x] данные о добавленных продуктах хранятся в localStorage и при перезагрузке страницы восстанавливаются `+5`
   - [x] добавление и сохранение данных в localStorage происходит также и с других страниц приложения `+5`
5. Реализован промокод блок +10
   - [x] реализовано поле ввода промокодов. При этом, если промокод найден во время ввода, то должна быть возможность добавления промокода в список примененных. При добавлении каждого промокода, зачеркивается старая итоговая цена на странице корзины(итоговая цена к оплате за все товары) и отображается новая(например под ней), с учетом скидки промокода. Если промокодов более одного, то суммируется их общая скидка и применяется к итоговой цене. Например, было применено 2 промокода, каждый из которых даёт 10% скидки. Суммируем и получаем скидку в размере 20%, затем применяем её к цене, которая образовалась до применения всех промокодов. Кол-во примененных промокодов не ограничено. Кроме того, должна быть возможность удаления промокода из списка примененных. Итоговая цена при этом также пересчитывается `+5`
   - [x] реализован блок примененных промокодов, где указаны все примененные промокоды и размер скидки, которые они дают, при этом цена до применения всех промокодов должна быть перечеркнута, и добавлена новая цена, с учетом скидок всех промокодов `+5`
6. Реализована кнопка открытия [модального окна оформления покупки](purchase-modal.md) + 5
   - [x] при клике на кнопку открывается модальное окно для оформления заказа `+5`
7. Реализован блок с общей суммой и кол-вом всех выбранных товаров +5
   - [x] блок отображает сумму и кол-во всех товаров и реагирует на применение промокодов (дополнительно указывается новая цена после применения промокодов) `+5`   
   
#### Модальное окно оформления товара  
  
  1. Реализован блок ввода персональной информации с валидацией +20
   - [x] добавлено поле "Имя и Фамилия". Валидация: содержит не менее двух слов, длина каждого не менее 3 символов `+5`
   - [x] добавлено поле "Номер телефона". Валидация: должно начинаться с '+', содержать только цифры и быть не короче 9 символов `+5`
   - [x] добавлено поле "Адрес доставки". Валидация: содержит не менее трех слов, длина каждого не менее 5 символов `+5`
   - [x] добавлено поле "E-mail". Валидация: проверяется, является ли введенный текст электронной почтой `+5`
2. Реализован блок ввода данных банковской карты с валидацией +20
   - [x] реализован ввод номер карты. Валидация: кол-во введенных цифр должно быть ровно 16, допускается ввод только цифр `+5`
   - [x] реализована автоматическая смена логотипа платежной системы. Например, если номер карты начинает с 4, устанавливается логотип Visa, если 5 - MasterCard. Реализовать не менее 3 платежных систем. `+5`
   - [x] реализован блок ввода срока действия карты. Валидация: допускается ввод только цифр, месяц не может быть больше 12, длина поля должна быть равна 4. Например 12/25. Разделитель не учитывается и добавляется автоматически `+5`
   - [x] реализован блок ввод CVV кода. Валидация: длина 3 символа, допускается ввод только цифр `+5`
3. Реализована кнопка завершения заказа +10
   - [x] при клике на кнопку submit/confirm проверяются все поля на валидность, если у поля есть ошибки валидации, то рядом с этим полем выводится сообщение с ошибкой `+5`
   - [x] при успешном прохождении валидации всех полей и нажатии на кнопку, выводится сообщение, что заказ оформлен. Затем, спустя 3-5 секунд происходит редирект на главную страницу магазина. Корзина при этом очищается `+5`  
   
#### Страница с описанием товара  
  
1. Реализованы блоки страницы +30
   - [x] присутствуют "хлебные крошки", указывающие на путь товара относительно корня сайта, например STORE > LAPTOPS > APPLE > MACBOOK PRO `+5`
   - [x] реализован блок с фотографиями товара, которые можно увеличивать при клике/наведении. Фотографии не должны содержать дубликаты/повторы `+10`
   - [x] реализован блок с полными данными товара (название, категория, описание, цена и т.д.) `+5`
   - [x] присутствует кнопка добавления товара в корзину. Состояние кнопки зависит от того, есть ли товар в корзине. Например, товар мог быть добавлен в корзину на текущей странице или [странице товаров](store-page.md) и соответственно текст кнопки должен это отображать `+5`
   - [x] присутствует кнопка быстрой покупки товара. При клике, если товара нет в корзине, происходит автоматическое добавление в корзину и переход на страницу корзины, с уже открытым модальным окном. Если товар уже был в корзине, повторное добавление не требуется `+5`
2. Страница открывается в новом окне по ссылке с id/name товара +10
   - [x] страница конкретного товара имеет свой маршрут. Например, страница с товаром [MacBook Pro](https://online-store-rs.netlify.app/product-details/6) открывается вне зависимости от того, посещались ли другие страницы приложения `+10`  
   
#### Header  
  
1. Header содержит корзину товаров +10
   - [x] присутствует корзина с количеством выбранных товаров `+5`
   - [x] количество выбранных товаров динамически изменяется при добавлении/удалении товара на всех страницах ([корзина](cart-page.md), [карточка товара](product-details-page.md), [поиск товаров](store-page.md)). А также включает кол-во одного вида товара, который может быть добавлен больше 1шт на [странице корзины](cart-page.md) `+5`
2. Header содержит общую сумму покупок +10
   - [x] присутствует корзина с суммой выбранных товаров `+5`
   - [x] сумма выбранных товаров динамически изменяется при добавлении/удалении товара на всех страницах ([корзина](cart-page.md), [карточка товара](product-details-page.md), [поиск товаров](store-page.md)). А также включает кол-во одного вида товара, который может быть добавлен больше 1шт на [странице корзины](cart-page.md) `+5`  
   
#### Страница 404  

1. Страница существует +6
   - [x] при переходе по несуществующему адресу открывается страница 404. Например, 'https://your-deploy.com/gfdgd' в случе если страницы с названием 'gfdgd' не существует `+6`
2. Страница не реагирует на некорректные query-параметры +4
   - [x] при появлении необрабатываемых(в вашем коде) query-параметров, страница не открывается. Например, 'https://your-deploy.com/?rgfg=dfoh'. Если в вашем случае параметр '?rgfg=dfoh' некорректен, то он не должен направлять на страницу 404 `+4`

MOONcitizenX committed:
style: fix reset button style

mkoroleva5 committed:
Merge pull request #62 from mkoroleva5/47-add-loaders-before-image-is-loaded

feat: add image beforeLoad spinner

mkoroleva5 committed:
Merge pull request #63 from mkoroleva5/61-add-reset-filters-button

61 add reset filters button

mkoroleva5 committed:
fix: change search input outline

mkoroleva5 committed:
fix: add 2 loader for both preview images

MOONcitizenX committed:
Merge pull request #69 from mkoroleva5/fix/styles-fixes

MOONcitizenX committed:
Merge pull request #68 from mkoroleva5/47-add-loaders-before-image-is-loaded

mkoroleva5 committed:
fix: add trim to search value in url

MOONcitizenX committed:
Merge pull request #70 from mkoroleva5/66-fix-search-input-add-trim-to-input-value

mkoroleva5 committed:
feat: add found matches to catalog

mkoroleva5 committed:
Merge branch 'develop' into 59-add-no-products-match-query-placeholder

mkoroleva5 committed:
fix: refactor matches function

mkoroleva5 committed:
refactor: change catalog/ filter tags to sections

mkoroleva5 committed:
fix: adjust matches padding

MOONcitizenX committed:
Merge pull request #71 from mkoroleva5/59-add-no-products-match-query-placeholder

59 add no products match query placeholder

mkoroleva5 committed:
feat: add quantity to filter options

MOONcitizenX committed:
Merge pull request #72 from mkoroleva5/48-add-filtered-products-quantity

feat: add quantity to filter options

MOONcitizenX committed:
fix: fix price min/max numbers in filter being integers

MOONcitizenX committed:
feat: add disabled filter label

MOONcitizenX committed:
fix: fix some disabled label styles

MOONcitizenX committed:
config: make tcm store types in separate folder and TS watch it

mkoroleva5 committed:
Merge pull request #76 from mkoroleva5/75-fix-price-range-being-integer-not-float

fix: fix price min/max numbers in filter being integers

mkoroleva5 committed:
Merge pull request #77 from mkoroleva5/74-add-disabled-state-to-filter-labels

feat: add disabled filter label

mkoroleva5 committed:
Merge pull request #79 from mkoroleva5/78-configure-typed-css-modules-and-ts-to-work-with-1-separate-folder

config: make tcm store types in separate folder and TS watch it

MOONcitizenX committed:
feat: add initial routing

MOONcitizenX committed:
Merge pull request #80 from mkoroleva5/37-create-router

feat: add initial routing

mkoroleva5 committed:
feat: add categories menu

MOONcitizenX committed:
Merge pull request #82 from mkoroleva5/81-add-categories-menu

feat: add categories menu

mkoroleva5 committed:
feat: add product page layout

mkoroleva5 committed:
feat: add images array in product card

mkoroleva5 committed:
feat: add link to brand in bread crumbs

mkoroleva5 committed:
fix: adjust images styles in product page

mkoroleva5 committed:
feat: add category and type with links

mkoroleva5 committed:
feat: add fast-purchase button

MOONcitizenX committed:
Merge pull request #85 from mkoroleva5/83-add-product-description-page

83 add product description page

MOONcitizenX committed:
feat: add LS helper functions

mkoroleva5 committed:
feat: add cart display

mkoroleva5 committed:
feat: add cart products generation

mkoroleva5 committed:
fix: change product card props

mkoroleva5 committed:
feat: create cart reducer

mkoroleva5 committed:
Merge pull request #89 from mkoroleva5/88-add-local-storage-helper-functions

feat: add LS helper functions

MOONcitizenX committed:
feat: add cart state

MOONcitizenX committed:
fix: separate image with spinners

MOONcitizenX committed:
refactor: move store files into store folder

MOONcitizenX committed:
feat; add productPage button dispatch

MOONcitizenX committed:
feat: add productPage button dispatch

MOONcitizenX committed:
feat: add local storage helper functions

MOONcitizenX committed:
Merge branch '87-create-cart-page-layout' of github.com:mkoroleva5/online-store into 87-create-cart-page-layout

mkoroleva5 committed:
Merge pull request #91 from mkoroleva5/87-create-cart-page-layout

87 create cart page layout

mkoroleva5 committed:
fix: remove hover state from ProductCard

mkoroleva5 committed:
Merge pull request #92 from mkoroleva5/90-remove-usestate-hover-from-productcard

fix: remove hover state from ProductCard

mkoroleva5 committed:
feat: add icon to delete button in cart

mkoroleva5 committed:
feat: add changing of button "In cart"

MOONcitizenX committed:
Merge pull request #94 from mkoroleva5/93-add-styles-to-delete-buttons-in-cart

MOONcitizenX committed:
Merge pull request #96 from mkoroleva5/95-change-adding-in-cart-button-when-product-added

mkoroleva5 committed:
feat: add copy link-button to filters

MOONcitizenX committed:
Merge pull request #97 from mkoroleva5/64-add-copy-link-button-to-filters

feat: add copy link-button to filters

MOONcitizenX committed:
feat: add name/price asc/desc sort to catalog

mkoroleva5 committed:
Merge pull request #100 from mkoroleva5/50-add-ascdesc-nameprice-sort-to-filter-state

feat: add name/price asc/desc sort to catalog

MOONcitizenX committed:
feat: add debounce to checkboxes

MOONcitizenX committed:
refactor: change sort from 'по алфавиту' to 'по названию'

MOONcitizenX committed:
Merge pull request #108 from mkoroleva5/50-add-ascdesc-nameprice-sort-to-filter-state

refactor: change sort from 'по алфавиту' to 'по названию' #50

MOONcitizenX committed:
feat: add cart writing it's state into local storage

MOONcitizenX committed:
fix: fix Product import

MOONcitizenX committed:
refactor: remove unwanted comments

MOONcitizenX committed:
feat: make initial cart state read local storage

MOONcitizenX committed:
fix: fix CartState context type

mkoroleva5 committed:
Merge pull request #106 from mkoroleva5/104-use-debounce-on-checkboxes

feat: add debounce to checkboxes

mkoroleva5 committed:
Merge pull request #110 from mkoroleva5/109-save-cart-state-to-local-storage

109 save cart state to local storage

mkoroleva5 committed:
Merge branch 'develop' into 112-fix-cartstate-context-type

mkoroleva5 committed:
Merge pull request #113 from mkoroleva5/112-fix-cartstate-context-type

fix: fix CartState context type

mkoroleva5 committed:
fix: adjust product card height

mkoroleva5 committed:
Merge pull request #116 from mkoroleva5/115-fix-product-cards-height

fix: adjust product card height

mkoroleva5 committed:
feat: add stock to products in cart

mkoroleva5 committed:
feat: add indexes to products in cart

mkoroleva5 committed:
feat: add sort by index in cart

mkoroleva5 committed:
refactor: rename product index in cart

MOONcitizenX committed:
refactor: separate basic components into separate files

mkoroleva5 committed:
feat: change gaining indexes of products in cart

mkoroleva5 committed:
Merge pull request #118 from mkoroleva5/111-separate-basic-components-into-files

refactor: separate basic components into separate files

mkoroleva5 committed:
Merge branch 'develop' into 101-add-more-info-to-products-in-cart

MOONcitizenX committed:
Merge pull request #117 from mkoroleva5/101-add-more-info-to-products-in-cart

mkoroleva5 committed:
fix: adjust spinner size in cart

mkoroleva5 committed:
feat: add pagination in cart

mkoroleva5 committed:
feat: create modal checkout window layout

MOONcitizenX committed:
fix: fix directing to previous page when products on last page are deleted

MOONcitizenX committed:
fix: fix cart pagination listen to query params

MOONcitizenX committed:
fix: fix last page not going -1 when deleting items from last page

mkoroleva5 committed:
Merge pull request #119 from mkoroleva5/107-create-cart-pagination

107 create cart pagination

mkoroleva5 committed:
Merge branch 'develop' into 120-create-modal-checkout-window

MOONcitizenX committed:
feat: add promocodes logic

mkoroleva5 committed:
feat: add styles to promos

MOONcitizenX committed:
Merge pull request #122 from mkoroleva5/121-add-promocodes

121 add promocodes

mkoroleva5 committed:
feat: change total cost in header and cart

mkoroleva5 committed:
Merge pull request #123 from mkoroleva5/121-add-promocodes

feat: change total cost in header and cart #121

mkoroleva5 committed:
feat: add closing modal function

mkoroleva5 committed:
feat: add phone number input mask

MOONcitizenX committed:
feat: add correct routing

MOONcitizenX committed:
feat: add 404 page

MOONcitizenX committed:
fix: fix 404 on 'cart' page

mkoroleva5 committed:
Merge pull request #125 from mkoroleva5/124-fix-routes-and-add-404-page

124 fix routes and add 404 page

mkoroleva5 committed:
Merge branch 'develop' into 120-create-modal-checkout-window

mkoroleva5 committed:
feat: add modal inputs to basic components

mkoroleva5 committed:
feat: add inputs masks, patterns, validation

mkoroleva5 committed:
refactor: delete inputs from basic components

mkoroleva5 committed:
feat: create order page and clear cart action

mkoroleva5 committed:
feat: add form submit validation and hints for inputs

mkoroleva5 committed:
feat: clear promos in CartState after order

mkoroleva5 committed:
feat: enlarge images in ProductPage by click

mkoroleva5 committed:
fix: adjust enlarged image role

mkoroleva5 committed:
feat: adjust spinner size in ProductPage

mkoroleva5 committed:
feat: change not found message

mkoroleva5 committed:
feat: add default sort by name

mkoroleva5 committed:
feat: add default sort by price down

mkoroleva5 committed:
fix: adjust 2 labels in FilterOption

MOONcitizenX committed:
Merge pull request #126 from mkoroleva5/120-create-modal-checkout-window

120 create modal checkout window

MOONcitizenX committed:
Merge pull request #130 from mkoroleva5/102-add-something-to-not-found-message

102 add something to not found message

MOONcitizenX committed:
Merge pull request #128 from mkoroleva5/127-add-image-enlargement-in-productpage

127 add image enlargement in productpage

MOONcitizenX committed:
Merge pull request #131 from mkoroleva5/98-fix-2-labels-in-filteroption-component

fix: adjust 2 labels in FilterOption

MOONcitizenX committed:
fix: fix dualsliders taking value from search params without min/max checking the value

mkoroleva5 committed:
fix: change DualSlider when other filters applied

mkoroleva5 committed:
Merge pull request #133 from mkoroleva5/103-fix-filter-ranges-when-changing-filters-and-query

103 fix filter ranges when changing filters and query

mkoroleva5 committed:
feat: add fast purchase action

MOONcitizenX committed:
Merge pull request #135 from mkoroleva5/134-add-fast-purchase-possibility

feat: add fast purchase action

mkoroleva5 committed:
feat: make filters adaptive

mkoroleva5 committed:
feat: make cart adaptive

mkoroleva5 committed:
fix: adjust links color in footer

mkoroleva5 committed:
feat: add favicon

mkoroleva5 committed:
feat: update data in products.json

mkoroleva5 committed:
feat: update data

mkoroleva5 committed:
feat: update data

mkoroleva5 committed:
feat: update data

MOONcitizenX committed:
Merge pull request #139 from mkoroleva5/22-update-datajson

MOONcitizenX committed:
Merge pull request #137 from mkoroleva5/136-make-filters-block-adaptive

mkoroleva5 committed:
fix: fix incorrect path

mkoroleva5 committed:
feat: add custom cursor to enlarged image wrapper

MOONcitizenX committed:
Merge pull request #138 from mkoroleva5/132-add-favicon

feat: add favicon

MOONcitizenX committed:
Merge pull request #141 from mkoroleva5/140-add-styles-to-enlarged-image-in-productcard

feat: add custom cursor to enlarged image wrapper

MOONcitizenX committed:
feat: add lazy load for catalog images

MOONcitizenX committed:
fix: remove isCheckout when order finished

MOONcitizenX committed:
feat: add BackToTop button to catalog and cart

MOONcitizenX committed:
feat: add lazyloading to cart

MOONcitizenX committed:
refactor: remove comments

mkoroleva5 committed:
Merge pull request #143 from mkoroleva5/142-add-lazy-loading-for-catalog-images

142 add lazy loading for catalog images

MOONcitizenX committed:
fix: fix product card img styles

mkoroleva5 committed:
Merge pull request #146 from mkoroleva5/142-add-lazy-loading-for-catalog-images

fix: fix product card img styles

mkoroleva5 committed:
Merge pull request #145 from mkoroleva5/144-add-backtotop-button-to-catalog

feat: add BackToTop button to catalog and cart

MOONcitizenX committed:
fix: fix BackToTop arrow color

mkoroleva5 committed:
Merge pull request #147 from mkoroleva5/144-add-backtotop-button-to-catalog

fix: fix BackToTop arrow color

mkoroleva5 committed:
fix: change div to link for cart in header

mkoroleva5 committed:
fix: adjust
line breaks in descriptions of products

mkoroleva5 committed:
fix: adjust cards width in cart

mkoroleva5 committed:
fix: fix modal closing when history.back

mkoroleva5 committed:
fix: change credit card background

MOONcitizenX committed:
Merge pull request #149 from mkoroleva5/148-fix-styles

148 fix styles

mkoroleva5 committed:
fix: fix bugs in header and footer

MOONcitizenX committed:
Merge pull request #150 from mkoroleva5/148-fix-styles

fix: fix bugs in header and footer

MOONcitizenX committed:
fix: try to fix netlify routing

MOONcitizenX committed:
Merge pull request #151 from mkoroleva5/148-fix-styles

fix: try to fix netlify routing

mkoroleva5 committed:
fix: fix arrow size in Safari in Filter

mkoroleva5 committed:
Merge branch '148-fix-styles' of https://github.com/mkoroleva5/online-store into 148-fix-styles

mkoroleva5 committed:
Merge pull request #152 from mkoroleva5/148-fix-styles

148 fix styles

mkoroleva5 committed:
fix: fix links color in CategoriesMenu

mkoroleva5 committed:
Merge pull request #153 from mkoroleva5/148-fix-styles

fix: fix links color in CategoriesMenu

mkoroleva5 committed:
fix: make filters title clickable on mobile

mkoroleva5 committed:
Merge pull request #154 from mkoroleva5/148-fix-styles

fix: make filters title clickable on mobile

qeteq reviewed (commented):
❗ One more thing: remove all empty `onKeyDown` handlers (or implement them).

MOONcitizenX committed:
fix: minor style fixes

MOONcitizenX committed:
feat: add chevron image to breadcrumbs

MOONcitizenX committed:
fix: fix product card styles

MOONcitizenX committed:
feat: add checkout modal window adaptive styles

mkoroleva5 committed:
Merge pull request #156 from mkoroleva5/155-fix-spinner-and-images-size-in-productcard

155 fix spinner and images size in productcard

mkoroleva5 committed:
fix: adjust enlarged image styles

mkoroleva5 committed:
fix: display copy/delete buttons when filter is open

mkoroleva5 committed:
fix: fix width of order page

mkoroleva5 committed:
Merge pull request #157 from mkoroleva5/148-fix-styles

148 fix styles

mkoroleva5 committed:
fix: fix outline of enlarged image

mkoroleva5 committed:
Merge pull request #158 from mkoroleva5/148-fix-styles

fix: fix outline of enlarged image

mkoroleva5 committed:
fix: change outline to box-shadow for enlarged image

mkoroleva5 committed:
Merge pull request #159 from mkoroleva5/148-fix-styles

fix: change outline to box-shadow for enlarged image

mkoroleva5 committed:
fix: fix cart reducer types

mkoroleva5 committed:
Merge pull request #160 from mkoroleva5/fix/fix-cart-reducer-types

fix: fix cart reducer types

mkoroleva5 committed:
fix: fix product page layout

mkoroleva5 committed:
Merge pull request #161 from mkoroleva5/148-fix-styles

fix: fix product page layout

MOONcitizenX committed:
fix: fix modal window scrollbar styles

mkoroleva5 committed:
Merge pull request #162 from mkoroleva5/148-fix-styles

fix: fix modal window scrollbar styles

MOONcitizenX committed:
fix: fix cart search params bugs

MOONcitizenX committed:
fix: fix data maslo maslo

MOONcitizenX committed:
fix: cover checkSearch with useCallback

MOONcitizenX committed:
fix: fix footer year to 2023

MOONcitizenX committed:
fix: remove data items with 0 stock

mkoroleva5 committed:
feat: replace jpg images with webp

mkoroleva5 committed:
Merge pull request #163 from mkoroleva5/148-fix-styles

148 fix styles

MOONcitizenX committed:
fix: fix cart pagination bug

MOONcitizenX committed:
fix: fix 'Address' field not accepting digits

MOONcitizenX committed:
refactor: remove console.log

mkoroleva5 committed:
Merge pull request #165 from mkoroleva5/164-fix-cart-pagination

fix: fix cart pagination bug

mkoroleva5 committed:
Merge pull request #167 from mkoroleva5/166-minor-fixes

fix: fix 'Address' field not accepting digits

MOONcitizenX committed:
feat: setup vitest configs

MOONcitizenX committed:
fix: fix cartState fileds being Objects

MOONcitizenX committed:
feat: add Footer tests

MOONcitizenX committed:
feat: add Header tests

MOONcitizenX committed:
refactor: rename some Header Tests

mkoroleva5 committed:
feat: write test for checkFilterState function

mkoroleva5 committed:
feat: write test for countItems function

mkoroleva5 committed:
feat: write test for formatPrice function

mkoroleva5 committed:
feat: write test for pathnameHelpers function

mkoroleva5 committed:
feat: write test for sortArray function

mkoroleva5 committed:
feat: write test for localStorageHelpers

MOONcitizenX committed:
fix: fix format price function

MOONcitizenX committed:
fix: fix format price tests

MOONcitizenX committed:
fix: fix useDebounce typing

MOONcitizenX commented on src/components/basic-components/BackToTop.tsx:
```diff
@@ -0,0 +1,43 @@
+  };
+
+  useEffect(() => {
+    window.addEventListener('scroll', listenerIsBackToTop);
```
Fixed

MOONcitizenX committed:
fix: fix 'passive' options on scroll events https://github.com/mkoroleva5/online-store/pull/65/files#r1064003918

MOONcitizenX commented on src/components/basic-components/DualSlider.tsx:
```diff
@@ -0,0 +1,101 @@
+  const sliderName = sliderGroup[0].toUpperCase() + sliderGroup.slice(1);
+  const filterState = useContext(FilterState);
+
+  let filterMin = sliderGroup === 'price' ? filterState.minPrice : filterState.minStock;
+  let filterMax = sliderGroup === 'price' ? filterState.maxPrice : filterState.maxStock;
+  if (filterMin && filterMax) {
+    if (filterMin > max && filterMax < min) {
+      filterMin = (filteredMax ?? filterMax ?? max) - 1;
+      filterMax = (filteredMin ?? filterMin ?? min) + 1;
+    }
+  }
+  if (filterMin) {
+    if (filterMin < min) filterMin = min;
+    else if (filterMin > max) filterMin = (filteredMax ?? filterMax ?? max) - 1;
+  }
+  if (filterMax) {
+    if (filterMax > max) filterMax = max;
+    else if (filterMax < min) filterMax = (filteredMin ?? filterMin ?? min) + 1;
+  }
```
Fixed

MOONcitizenX committed:
fix: move logic into separate function https://github.com/mkoroleva5/online-store/pull/65/files#r1063397361

MOONcitizenX commented on src/components/basic-components/FilterOption.tsx:
```diff
@@ -0,0 +1,70 @@
+        if (currentGroup) {
+          updateSearchValue(
+            filterGroupString,
+            [...currentGroup.split('-and-'), searchValue].join('-and-'),
```
Fixed

MOONcitizenX committed:
fix: extract split/join logic into separate function https://github.com/mkoroleva5/online-store/pull/65/files#r1060825369

MOONcitizenX commented on src/components/basic-components/ImageSpinner.tsx:
```diff
@@ -0,0 +1,20 @@
+interface ImageSpinnerProps {
+  displayList?: boolean;
+  sizeClass?: string;
+  classN?: string;
```
Fixed

MOONcitizenX committed:
fix: fix ImageSpinner props naming and interface
https://github.com/mkoroleva5/online-store/pull/65/files#r1063374613

MOONcitizenX committed:
fix: use cartState for checks in 'InCartButton'
https://github.com/mkoroleva5/online-store/pull/65/files#r1060830635

MOONcitizenX committed:
fix: fix spinner enlarged issues

MOONcitizenX commented on src/components/basic-components/InCartButton.tsx:
```diff
@@ -0,0 +1,30 @@
+        e.stopPropagation();
+        dispatch({ type: 'ADD_PRODUCT', payload: product });
+        const target = e.target as HTMLElement;
+        if (target.textContent === 'В корзине') history.push('/cart');
```
Fixed

MOONcitizenX committed:
refactor: extract repeating dispatch correct promo logic into separate function
https://github.com/mkoroleva5/online-store/pull/65/files#r1060833271

MOONcitizenX commented on src/components/basic-components/PromoCodes.tsx:
```diff
@@ -0,0 +1,81 @@
+          disabled={!Object.keys(possiblePromos).includes(input)}
+          type="button"
+          onClick={() => {
+            if (Object.keys(possiblePromos).includes(input)) {
+              dispatch({
+                type: 'ADD_PROMO',
+                payload: input,
+              });
+              setInput('');
+            }
```
Fixed

MOONcitizenX commented on src/components/cartState.tsx:
```diff
@@ -0,0 +1,32 @@
+  WINTER: 15,
+};
+
+export interface CartStateProps {
```
Fixed

MOONcitizenX commented on src/components/cartState.tsx:
```diff
@@ -0,0 +1,32 @@
+  promos: savedCartState?.promos ?? [],
+};
+
+export const CartState = createContext<CartContext>({} as CartContext);
```
Fixed

MOONcitizenX committed:
refactor: rename CartState interface and CartStateContext to be meaningful
https://github.com/mkoroleva5/online-store/pull/65/files#r1060793445

MOONcitizenX commented on src/components/catalog/Catalog.tsx:
```diff
@@ -0,0 +1,182 @@
+            <Page404 />
+          ) : (
+            <>
+              {isCartOpen && <Cart />}
+              {!isCartOpen && isProductPageView && product && <ProductPage product={product} />}
+              {!isCartOpen && !isProductPageView && (
```
Fixed

MOONcitizenX committed:
refactor: extract jsx conditional rendering logic outside jsx
https://github.com/mkoroleva5/online-store/pull/65/files#r1063416322

MOONcitizenX commented on src/components/catalog/cart/Cart.tsx:
```diff
@@ -0,0 +1,203 @@
+  });
+
+  useEffect(() => {
+    const pageQ = getSearchValue('page');
```
Query

MOONcitizenX commented on src/components/catalog/cart/Cart.tsx:
```diff
@@ -0,0 +1,203 @@
+      updateSearchValue('limit', '3');
+    }
+
+    const unlisten = history.listen(() => {
```
Fixed

MOONcitizenX committed:
fix: add useEffect remove listener
https://github.com/mkoroleva5/online-store/pull/65/files#r1063417798

MOONcitizenX committed:
refactor: useEffect return statements without unnecessary arrow functions

MOONcitizenX commented on src/components/catalog/cart/CheckoutPage.tsx:
```diff
@@ -89,6 +89,17 @@ export const CheckoutPage = () => {
+        dispatch({ type: 'SET_CHECKOUT', payload: false });
+      }
+    });
+    return () => {
+      unlisten();
+    };
```
Fixed

MOONcitizenX committed:
refactor: move product filteredStock/maxStock count logic into separate constants

MOONcitizenX commented on src/components/catalog/filter/Filter.tsx:
```diff
@@ -0,0 +1,134 @@
+                  value={product}
+                  id={index}
+                  filterGroup="product"
+                  quantity={`${filteredProducts.filter((el) => el.type === product).length}/${
+                    products.filter((el) => el.type === product).length
```
Fixed

MOONcitizenX committed:
refactor: change 'classN' to 'imageSize' for better understanding
https://github.com/mkoroleva5/online-store/pull/65/files#r1063372618

MOONcitizenX commented on src/components/catalog/productPage/ProductPageImage.tsx:
```diff
@@ -0,0 +1,41 @@
+  src: string;
+  title: string;
+  index?: number;
+  classN: 'fullImg' | 'img' | 'enlargedImg';
```
Fixed

MOONcitizenX commented on src/components/catalog/sortMenu/CatalogMenu.tsx:
```diff
@@ -0,0 +1,93 @@
+            [style.sortBtnActive]: filterState.sort === 'namedown' || filterState.sort === 'nameup',
+          })}
+          onClick={() => {
+            updateSearchValue('sort', getSearchValue('sort') === 'nameup' ? 'namedown' : 'nameup');
```
Fixed

MOONcitizenX commented on src/components/catalog/sortMenu/CatalogMenu.tsx:
```diff
@@ -0,0 +1,93 @@
+import { getSearchValue, updateSearchValue } from '../../../utils/searchHelpers';
+
+export const CatalogMenu = () => {
+  const filterState = useContext(FilterState);
```
Fixed

MOONcitizenX committed:
refactor: use destructuring of state
https://github.com/mkoroleva5/online-store/pull/65/files#r1063993817
and get 'sort' value directly from state
https://github.com/mkoroleva5/online-store/pull/65/files#r1063993873

MOONcitizenX committed:
fix: remove unused import

mkoroleva5 committed:
fix: make countItems function separate https://github.com/mkoroleva5/online-store/pull/65/files/5536210a39533f7bff8f6007a7ca49f9ba80058d#r1060878207

mkoroleva5 commented on src/components/catalog/matches/Matches.tsx:
```diff
@@ -0,0 +1,16 @@
+
+export const Matches = ({ length }: MatchesProps) => {
+  const lastNum = +length.toString().split('').reverse()[0];
+  const found = lastNum === 1 && length !== 11 ? 'Найден' : 'Найдено';
```
Fixed

mkoroleva5 committed:
Merge branch '169-refactoring' of https://github.com/mkoroleva5/online-store into 169-refactoring

MOONcitizenX commented on src/components/header/Header.module.css:
```diff
@@ -0,0 +1,114 @@
+  align-items: center;
+  gap: 15px;
+  text-decoration: none;
+  color: #333333;
```
Fixed

MOONcitizenX committed:
refactor: add one more var for colors and change it in css files
https://github.com/mkoroleva5/online-store/pull/65/files#r1060807308

MOONcitizenX committed:
Merge branch '169-refactoring' of github.com:mkoroleva5/online-store into 169-refactoring

mkoroleva5 committed:
refactor: refactor styles for display table/list in catalog https://github.com/mkoroleva5/online-store/pull/65/files/5536210a39533f7bff8f6007a7ca49f9ba80058d#r1063998187

mkoroleva5 commented on src/components/catalog/productCard/ProductCard.tsx:
```diff
@@ -0,0 +1,84 @@
+export const ProductCard = ({ product, layout, path }: ProductProps) => {
+  const [isImageLoaded, setIsImageLoaded] = useState(false);
+
+  const tableLayout = layout === 'table';
+  const listLayout = layout === 'list';
```
Fixed

mkoroleva5 committed:
refactor: remove comments from products.ts https://github.com/mkoroleva5/online-store/pull/65/files/5536210a39533f7bff8f6007a7ca49f9ba80058d#r1060807707

mkoroleva5 commented on src/data/product.ts:
```diff
@@ -0,0 +1,24 @@
+  amount: number;
+}
+
+// export type CategoryPaths = 'sweets' | 'candles' | 'drinks';
+
+// export type ProductCategory = 'Полезные сладости' | 'Полезные напитки' | 'Свечи';
+
+// export type ProductType = 'батончик' | 'печенье';
```
Fixed

mkoroleva5 commented on src/utils/countItems.ts:
```diff
@@ -0,0 +1,10 @@
+export const countItems = (length: number) => {
+  const lastNum = +length.toString().split('').reverse()[0];
```
Fixed

mkoroleva5 committed:
fix: change countTotalDiscount returned value to number https://github.com/mkoroleva5/online-store/pull/65/files/5536210a39533f7bff8f6007a7ca49f9ba80058d#r1064004284

mkoroleva5 commented on src/store/CartStore.ts:
```diff
@@ -0,0 +1,121 @@
+export const countTotalItems = (productsObj: Record<number, CartProduct>) => {
+  return Object.values(productsObj).reduce((acc, item) => acc + item.amount, 0);
+};
+export const countTotalCost = (productsObj: Record<number, CartProduct>) => {
+  return Object.values(productsObj)
+    .reduce((acc, item) => acc + item.amount * item.price, 0)
+    .toFixed(2);
+};
+
+export const countTotalCostDiscount = (totalCost: number, totalDiscount: number) => {
+  return (totalCost - (totalCost / 100) * totalDiscount).toFixed(2);
+};
+
```
Fixed

mkoroleva5 committed:
fix: remove unnecessary condition in removeLS function https://github.com/mkoroleva5/online-store/pull/65/files/5536210a39533f7bff8f6007a7ca49f9ba80058d#r1060874569

mkoroleva5 commented on src/utils/localStorageHelpers.ts:
```diff
@@ -0,0 +1,19 @@
+};
+
+export const removeLS = (key: string) => {
+  if (localStorage.getItem(key)) {
```
Fixed

mkoroleva5 committed:
fix: add try/catch to JSON.parse in getLS function https://github.com/mkoroleva5/online-store/pull/65/files/5536210a39533f7bff8f6007a7ca49f9ba80058d#r1060873884

mkoroleva5 committed:
feat: create Link component https://github.com/mkoroleva5/online-store/pull/65/files/5536210a39533f7bff8f6007a7ca49f9ba80058d#r1060804151

mkoroleva5 commented on src/components/header/Header.tsx:
```diff
@@ -0,0 +1,54 @@
+  return (
+    <header className={style.headerWrapper}>
+      <div className={style.header}>
+        <a
+          href="/"
+          className={style.logoWrapper}
+          onClick={(e) => {
+            e.preventDefault();
+            history.push('/');
+          }}
+        >
```
Fixed

mkoroleva5 committed:
fix: adjust formatPrice and countCost functions https://github.com/mkoroleva5/online-store/pull/65/files#r1060805549

mkoroleva5 commented on src/components/header/Header.tsx:
```diff
@@ -0,0 +1,54 @@
+  const totalCost = countTotalCost(cartState.products);
+  const totalItems = countTotalItems(cartState.products);
+  const totalCostDiscounted = countTotalCostDiscount(
+    +totalCost,
```
Fixed

mkoroleva5 commented on src/utils/localStorageHelpers.ts:
```diff
@@ -0,0 +1,19 @@
+export const getLS = (key: string) => {
+  const value = localStorage.getItem(key);
+  if (value) {
+    return JSON.parse(value);
```
Fixed

mkoroleva5 committed:
fix: prevent side effects in cartReducer https://github.com/mkoroleva5/online-store/pull/65/files#r1062705248

mkoroleva5 commented on src/store/CartStore.ts:
```diff
@@ -0,0 +1,120 @@
+          },
+        },
+      };
+      setLS(localStorageCartStateName, newState);
```
Fixed

MOONcitizenX committed:
feat: add searchHelpers tests

mkoroleva5 committed:
Merge pull request #171 from mkoroleva5/168-implement-tests

168 implement tests

mkoroleva5 committed:
Merge branch 'develop' into 169-refactoring

mkoroleva5 committed:
Merge pull request #170 from mkoroleva5/169-refactoring

169 refactoring

MOONcitizenX committed:
fix: fix imports in tests

MOONcitizenX committed:
docs: add README

MOONcitizenX committed:
Merge pull request #172 from mkoroleva5/169-refactoring

fix: fix imports in tests

MOONcitizenX committed:
Merge pull request #174 from mkoroleva5/173-add-readme

docs: add README

MOONcitizenX committed:
Update README.md

MOONcitizenX committed:
docs: update README

MOONcitizenX committed:
refactor: remove RS logo from footer

MOONcitizenX committed:
Merge pull request #175 from mkoroleva5/173-add-readme

refactor: remove RS logo from footer
